qutip/tests/test_bofin_solvers.py

```diff
@@ -0,0 +1,900 @@
+
+        eigvalsAP = eigvalsh(AlphaP)
+
+        chi = []
+        for val in eigvalsAP[0:lmax - 1]:
+            chi.append(-2/val)
+
+        eta_list = [
+            0.5 * lmax * (2 * (lmax + 1) - 1) * (
+                np.prod([chi[k]**2 - eps[j]**2 for k in range(lmax - 1)]) /
+                np.prod([
+                    eps[k]**2 - eps[j]**2 + deltafun(j, k) for k in range(lmax)
+                ])
+            )
+            for j in range(lmax)
+        ]
+
+        kappa = [0] + eta_list
+        epsilon = [0] + eps
+
+        def f_approx(x):
+            f = 0.5
+            for ll in range(1, lmax + 1):
+                f = f - 2 * kappa[ll] * x / (x**2 + epsilon[ll]**2)
+            return f
+
+        def C(sigma, mu):
+            eta_0 = 0.5 * Gamma * W * f_approx(1.0j * beta * W)
+            gamma_0 = W - sigma*1.0j*mu
+            eta_list = [eta_0]
+            gamma_list = [gamma_0]
+            if lmax > 0:
+                for ll in range(1, lmax + 1):
+                    eta_list.append(
+                        -1.0j * (kappa[ll] / beta) * Gamma * W**2
+                        / (-(epsilon[ll]**2 / beta**2) + W**2)
+                    )
+                    gamma_list.append(epsilon[ll]/beta - sigma*1.0j*mu)
+            return eta_list, gamma_list
+
+        etapL, gampL = C(1.0, mu_l)
+        etamL, gammL = C(-1.0, mu_l)
+
+        etapR, gampR = C(1.0, mu_r)
+        etamR, gammR = C(-1.0, mu_r)
+
+        ck_plus = etapR + etapL
+        vk_plus = gampR + gampL
+        ck_minus = etamR + etamL
+        vk_minus = gammR + gammL
+
+        return ck_plus, vk_plus, ck_minus, vk_minus
+
+
+class TestFermionicHEOMSolver:
+    @pytest.mark.filterwarnings("ignore::scipy.integrate.IntegrationWarning")
+    @pytest.mark.parametrize(['evo'], [
+        pytest.param("qobj"),
+        pytest.param("qobjevo"),
+        pytest.param("listevo"),
+    ])
+    @pytest.mark.parametrize(['liouvillianize'], [
+        pytest.param(False, id="hamiltonian"),
+        pytest.param(True, id="liouvillian"),
+    ])
+    def test_discrete_level_model(self, evo, liouvillianize, atol=1e-3):
+        dlm = DiscreteLevelCurrentModel(
+            gamma=0.01, W=1, T=0.025851991, lmax=10,
+        )
+        H_sys = hamiltonian_to_sys(dlm.H, evo, liouvillianize)
+        ck_plus, vk_plus, ck_minus, vk_minus = dlm.bath_coefficients()
+
+        options = Options(
+            nsteps=15_000, store_states=True, rtol=1e-14, atol=1e-14,
+        )
+        Ncc = 2  # For a single impurity we converge with Ncc = 2
+        hsolver = FermionicHEOMSolver(
+            H_sys, dlm.Q, ck_plus, vk_plus, ck_minus, vk_minus, Ncc,
+            options=options,
+        )
+
+        tlist = [0, 10]
+        result = hsolver.run(dlm.rho(), tlist, ado_return=True)
+        current = dlm.state_current(result.ado_states[-1])
+        analytic_current = dlm.analytic_current()
+        np.testing.assert_allclose(analytic_current, current, atol=atol)
```

Comment:
these tests are passing despite some bugs because atol is too high, and these currents are very small (order 1e-5), so maybe setting an rtol insted of atol is better.  Also, due to me messing up the directionality of the current, we will need a minus sign on one of them!

Reply:
Setting rtol would be better here. I'll do that as soon as we've figured out the correct current results so that I know what to set rtol to.

Reply:
Results corrected & tolerances fixed.